melview.py

```diff
@@ -185,6 +194,8 @@ def _refresh_device_info(self, retry=True):
                 self._otemp_list.append(float(self._json['outdoortemp']))
                 # Keep only last 10 temperature values.
                 self._otemp_list = self._otemp_list[-10:]
+            if 'zones' in self._json:
+                self._zones = self._json['zones']
```

Comment:
Just for my reference, I'm guessing this is just a list of objects like:
```
[
    {
        "zoneid": <int?>,
        "name": <str?>,
        "status": <int?>
    }
]
```

Reply:
If so we should change it into a dictionary so we don't need to iterate all zones for a given `zoneid`.

```python
{z['zoneid'] : z for z in self._json['zones']}
```

Reply:
Yes, correct. 
I originally had it like you have so I didn't have to iterate, but decided against it to keep the code looking a bit nicer. Iterating over a few items is very cheap.

Reply:
The changes here are to make grabbing specific zones simpler - you have a dictionary for lookups instead.

Currently that process would take a list of dictionary objects, encapsulate them all in classes, and then search throw the list of classes until you reach the correct zone - a huge waste given how simple the data structure is.

Reply:
yeah look happy to change this one, but I would still like the zone class. 
so it would basically be a dict of `zoneid -> MelViewZone`

Reply:
If you can justify the use of the class beyond simply encapsulating I might consider it - but as is I'm not keen to merge it.

Reply:
can I understand what is the disadvantage of the separation of concerns provided by the encapsulation?
Another advantage of it is providing named, typed values for the json.
Isn't it really hard to keep going back to the json response to see what kind of data you have and what kind of types are there?

As the project gets larger, it'll get more difficult to manage.
Isn't it nice to have all this json stuff abstracted away?

Reply:
I was mainly concerned with the instantiation during generation before. 

I definitely can appreciate the point of type hinting - that's something I do want to do eventually that I hadn't actually considered here.

I think it might be nicer then to bind the on/off methods onto the zone class too, in that case.

Reply:
To clarify, you mean the zone class will keep a reference to the parent `MelViewDevice` correct? and those functions will call `send_command` directly?

Reply:
Use `functools.wrap` to bind the on and off methods for zones to functions on the MelViewZone class.

Reply:
Oh wow I didn't know this was possible in python -- last time I touched python was years ago.
Can you walk me through how to do it?

Reply:
Sure - basically just pass `self` (the main MelView instance) to the zone init as a parameter and use something like:

```python
self.enable = functools.wrap(parentMelview.enable_zone, zoneid=self.zoneid)
```

Sorry for the brevity - I'm on my phone. 
```

Reply:
nice, thanks! where would you call `zone.enable()`? In `switch.py`? Because that already has the melView instance anyway because it calls `refresh` on it when it needs to update

Reply:
Woops I'd forgotten about that.

For now implement it with the binding plus passing the refresh as you currently do.
When I unify the setup stuff I'll have a single MelView instance and probably handle the platform updates independent of the components.

Reply:
sorry mate I'll have to come back to this later. I've pushed all the changes I've got so far so check them out!